48_ima_incremental_attestation.md

```diff
@@ -0,0 +1,340 @@
+notes and implementing the feature itself. Reviewers
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+The IMA event logs can grow to multiple thousands of log entries with each
+entry being ~100 bytes or even larger when IMA appraisal is being used on
+a system. With Keylime initiating frequent attestations, Mbytes of network bandwidth
+are consumed for the transfer of the IMA log and many CPU cycles are consumed
+by the verifier to verify each IMA log. This enhancement attempts to address
+the resource consumption issues by introducing incremental IMA attestation
+where the verifier requests only the difference in the IMA log (since last
+attestation) using the next-to-request IMA log entry number. This allows the
+verifier to resume the verification of the IMA log where it left off the
+previous time. For this to work, it also needs to maintain the state of the
+IMA PCR alongside the next-to-request IMA log entry number.
+
+In the ideal case, a verifier would only perform the check of the TPM quote
+rather than verifying the entire IMA log of a host that has not had any
+new entries in its IMA log.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this enhancement.  Describe why the change is important and the benefits to users.
+-->
+
+The motivation of this enhancement is driven by the realization that bandwidth
+consumption as well as the CPU cycle consumption can be greatly reduced and
+the scalability and efficiency of the Keylime verifier be increased when only
+the differences in the IMA logs from the last (previous) attestation verification are
+transferred.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+The goal of this enhancement is to reduce network bandwidth consumption as well as
+the verifier's CPU cycle consumption and increase overall scalability of
+Keylime when IMA attestation is being used.
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The verifier will need to maintain per-agent state that reflects the state
+of the IMA PCR from the last attestation along with the next-to-request
+IMA log entry. The verifier needs to be extended to request the
+IMA log starting at a specific entry in the IMA log. There are no changes
+required to the TPM quote.
+
+<!-- This proposal assumes that affinity of an agent with a verifier is
```

Comment:
The assumption that a given `agent` has strong affinity to a `verifier` is well warranted.

Reply:
It's not as severe anymore as it was before since we could theoretically read the last state of the attestation from the DB every time we receive a quote response. However, I am wondering whether any such setups actually exist today where there is a load balancer involved and what it actually does to a quote response due to the pull model that's implemented.

I can re-activate the text above that I commented out, and probably that would be better.

Reply:
No currently it's not possible to run verifiers behind a loadbalancer because of the verifier affinity (in-memory state). I just mentioned it because if we move to a push model instead we could remove that in-memory state and scale verifiers more easily behind LBs and get free HA as well.

Reply:
A load balancer should be feasible with a typical database.  The DB can either be accessed by the various verifiers, or the DB itself can also be load balanced / distributed.